docs/proposals-accepted/202206-block-series-deletion-compactor.md

```diff
@@ -0,0 +1,210 @@
+    mtx         sync.RWMutex
+}
+
+// Result for the processed result for each block.
+// Key: global tombstone ID, Value: processed result for each tombstone.
+type BlockTombstones map[ulid.ULID]*MemTombstones
+```
+
+Since all the blocks and tombstones are immutable, this result can be easily cached in memory or in filesystem. Ideally for each tombstone, each block we just need to calculate once. Since `BlockTombstones` is a map, it is easy to add or delete one key if users want to add more tombstones or remove one tombstone.
+
+During the query time, the cached tombstone will be used. As it knows which `SeriesRef`, which intervals to apply, it is very straightforward to mask series.
+
+Pros:
+* More efficient during query time, less processing required.
+* As data can be cached there is not much computation required.
+
+Cons:
+* More memory required.
+* If read is much less than write, it is inefficient as the pre-calculated result won't be used much.
+
+#### Considerations
+
+Considering the infrequent change of tombstones, option 2 is preferred as all data can be cached and it is faster during query time. 
+
+### Tombstone Deletion with Compactor
+
+Compactor performs the job of deleting the series from object storage and cleans up tombstone files.
+
+The main challenge is about scalability. The reasons are:
+* Global tombstone files makes the compactor difficult to know which blocks match the tombstones. It needs to scan all the block indexes to find the matched blocks.
+* If one block has too many matched series to mask, the query performance can be hugely impacted. In this case, compactor planner should be able to know the ratio of the masked series / total series. If the ratio is high, the block needs to be compacted and masked series should be removed from the block.
+
+There are other questions about how we should perform the deletion along with compaction:
+* Series are deleted at normal compaction. This might need to wait a long time and hard to apply for blocks at the max compaction level.
+* Rewrite blocks separately. This needs more coordination with compaction to avoid duplicate/unnecessary work.
+
+To overcome those challenges, we propose two options.
+
+#### Option 1: Compactor Scans Blocks
+
+Like store gateway, compactor can also scan block indexes to find the blocks matching the tombstones. Similar to the `Mask Series At Tombstone Sync Time` of store gateway, compactor can calculate the processed tombstones for each block, each tombstone and cache the results locally. As blocks and tombstones are immutable, the result needs to be calculated only once and then reused easily. As the Prometheus format tombstones are calculated, during the compaction time the compactor just needs to generate the tombstone file in the filesystem and the TSDB compactor will apply the deletion.
+
+There are two approaches about how to scan block indexes:
+1. Download all the blocks locally and check against the tombstones.
```

Comment:
We cannot download all blocks locally - we might not have enough disk for this. Should we instead scan one by one periodically? 

Would be nice to just think now how often we sync and what's the pseudoalgorithm for this. This would give us some idea about cost and delays - and tell us if we can start with option 1 or we should move to option 2 straight away 🤗